### PR TITLE
[REF] account, hr_{expense|holidays}, mail: move main attachment to mixin

### PR DESCRIPTION
**Pull request by pyduf**
Main attachment is quite costly and as it is implemented in mail.thread and is
therefore present in most model. In order to improve performance, we remove it
here from mail.thread and add it through a new mixin,only on model that
effectively use that features.

See sub commits for more details.

Task-2648976

### CODE REVIEW

File: addons/mail/models/mail_thread_main_attachment.py
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread_main_attachment.py:31]: While we are at it, this ``sudo`` is annoying me since I saw it ... but indeed it seems you can have the right to post on a document without having the write access (e.g. mail_post_access set to read, or replying to a notification on a document you can't read). It comes from https://github.com/odoo/odoo/pull/30659 .

I have a challenge (entrepreneur version) or a request (human version) for you: could you write a small unit test for this sudo ? I think we already have either a record with mail_post_access tweak, either an example of patched ACL in some mail tests (sorry for not being more precise, have to dig into those files :D ) . That would be great :)

File: addons/test_mail/tests/test_message_post.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_post.py:51]: As it is used in a single test can't we move it directly into the test itself ?

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for the quick PR :) not much to say, one small comment and one "not completely related" request :) .

Cheers !

File: addons/mail/models/mail_thread_main_attachment.py
Comment (pyduf) [reply addons/mail/models/mail_thread_main_attachment.py:31]: As discussed, it has been moved to a child task (3178885) for not stalling this one, see odoo/odoo#112273.

File: addons/test_mail/tests/test_message_post.py
Comment (pyduf) [reply addons/test_mail/tests/test_message_post.py]: Yes good idea , thanks.

**Comment by tde-banana-odoo** [comment]
@pyduf speaking of inheritance, I tried to fix the inheritance issue with https://github.com/odoo/odoo/pull/112095 doesn't that solve (part of) your inherit issue ?

**Comment by pyduf** [comment]
> @pyduf speaking of inheritance, I tried to fix the inheritance issue with #112095 doesn't that solve (part of) your inherit issue ?
> 
> Edit: just look at portal code, other commits are backported tests :D

Thanks for pointing me to that PR, it was indeed that method (not being called) that revealed the inheritance problem in  odoo/odoo#103966. I will see if it solves the problem without the portal-rating mixin.

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Added some suggestions & comments, nice work!

Cheers!

**Comment by pyduf** [comment]
> Hello @pyduf
> 
> Added some suggestions & comments, nice work!
> 
> Cheers!

Thanks for the review.

File: addons/mail/models/mail_thread_main_attachment.py
Comment (awa-odoo) [inline addons/mail/models/mail_thread_main_attachment.py:26]: Would add a comment to clarify it takes the first pdf if available, then the next image if available, then the first attachment

**Review by awa-odoo — COMMENTED** [review]
Hello @pyduf 

Looks good to me except for a minor comment.

Maybe we don't need that many commits though, the "CLN" ones could be squashed into a single "IMP" (I don't think we use CLN ?).

@tde-banana-odoo maybe you want to have a final look before merging?

Thanks for the follow-up and cheers!

File: addons/mail/models/mail_thread_main_attachment.py
Comment (pyduf) [reply addons/mail/models/mail_thread_main_attachment.py]: Yes indeed, good idea

**Comment by pyduf** [comment]
> Hello @pyduf
> 
> Looks good to me except for a minor comment.
> 
> Maybe we don't need that many commits though, the "CLN" ones could be squashed into a single "IMP" (I don't think we use CLN ?).
> 
> @tde-banana-odoo maybe you want to have a final look before merging?
> 
> Thanks for the follow-up and cheers!

I have checked and indeed, there are no CLN in the guidelines. Actually, I saw it on another PR. I have followed your advice and squashed all CLN into one IMP commit.

File: addons/test_mail/tests/test_message_post.py
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_post.py:932]: Would update test function name to hilight we are testing the main attachment mixin. We can also add a small docstring while we are at it :)

File: addons/mail/models/ir_attachment.py
Comment (tde-banana-odoo) [inline addons/mail/models/ir_attachment.py:25]: Could be great to expand a bit the commit message, telling it is code cleaning done during move of main attachment into its own mixin, that no functional changes should occur with this commit, that we mainly do pre-check for early returns, that some variable renaming is done to ease understanding, ...

**Review by tde-banana-odoo — APPROVED** [review]
LGTM, two side remarks while reading the final diff, thanks everyone here:)

File: addons/test_mail/tests/test_message_post.py
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_message_post.py:938]: As main attachment does the base job + some additional job, we can probably consider code is sufficiently tested.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Pull request closed without merging**
